### ISSUE TITLE
optimize PPO-training

### ISSUE THREAD

**Issue by tabee** [tabee/Reinforcement-Learning-Snake#27]
### **Strategie zur Optimierung der Timesteps für PPO-Training bei Snake**

Um das Training effizient zu gestalten und sicherzustellen, dass das Modell optimal trainiert wird, kannst du eine **progressive Trainingsstrategie** nutzen:

---

### **1. Modell regelmäßig speichern (Checkpoints)**
✅ **Verwende `save()`**, um das Modell in regelmäßigen Abständen zu speichern.  
Damit kannst du verschiedene Versionen später vergleichen und bei Bedarf das beste Modell laden.

#### **Code für regelmäßiges Speichern (alle 50k Timesteps)**:
```python
from stable_baselines3.common.callbacks import CheckpointCallback

checkpoint_callback = CheckpointCallback(
    save_freq=50000,  # Speichert alle 50k Timesteps
    save_path="./models/",
    name_prefix="ppo_snake"
)

model.learn(total_timesteps=500000, callback=checkpoint_callback)
```
📌 **Erklärung:**  
- Speichert alle **50'000 Timesteps** ein Modell in `./models/ppo_snake_50000.zip`, `ppo_snake_100000.zip`, etc.  
- So kannst du das Training später auswerten und das beste Modell auswählen.

---

### **2. TensorBoard zur Performance-Überwachung**
✅ Falls du TensorBoard nutzt, kannst du damit **Rewards, Losses und Entropy** beobachten.

#### **TensorBoard starten (falls nicht bereits geschehen):**
```bash
tensorboard --logdir=./tensorboard/
```

Achte besonders auf:
- **`ep_rew_mean`** (Durchschnittlicher Episoden-Reward) → Sollte kontinuierlich steigen.  
- **`policy_loss`** → Sollte nicht zu groß fluktuieren.  
- **`entropy_loss`** → Sollte nicht zu schnell gegen 0 gehen (sonst fehlt Exploration).  

Falls der Reward nach einer gewissen Anzahl Timesteps **nicht mehr steigt oder sinkt**, könnte **Overfitting** oder ein **lokales Optimum** vorliegen. Dann kannst du:
1. **Neues Training starten mit kleineren `clip_range` (z. B. 0.1 statt 0.2)**  
2. **Exploration durch `ent_coef=0.02` statt `0.01` verbessern**  

---

### **3. Bestes Modell automatisch auswählen**
✅ Falls du immer das **beste Modell** basierend auf dem höchsten Reward behalten willst, kannst du den **Best Model Saver** nutzen.

#### **Callback für Best Model Saver**
```python
from stable_baselines3.common.callbacks import EvalCallback

eval_callback = EvalCallback(
    env,
    best_model_save_path="./best_model/",
    log_path="./logs/",
    eval_freq=10000,  # Alle 10k Timesteps evaluieren
    deterministic=True,
    render=False
)

model.learn(total_timesteps=500000, callback=eval_callback)
```
📌 **Erklärung:**  
- Alle **10k Timesteps** wird das Modell getestet.  
- Falls der Durchschnittsreward das beste bisherige Ergebnis übertrifft, wird das Modell **automatisch gespeichert** (`./best_model/best_model.zip`).  
- So musst du nicht manuell vergleichen.

---

### **4. Training nach Pausen wieder aufnehmen**
Falls du das Training später fortsetzen willst, kannst du einfach das beste Modell laden:

```python
from stable_baselines3 import PPO

model = PPO.load("./best_model/best_model.zip", env=env)
model.learn(total_timesteps=500000)  # Training fortsetzen
```

---

### **5. Wann sollte das Training gestoppt werden?**
Nutze eine dieser **3 Methoden**, um den optimalen Punkt zum Stoppen zu bestimmen:

✅ **Methode 1: Reward-Stagnation**
- Falls der durchschnittliche Reward nach **100k Timesteps** nicht mehr steigt (z. B. für 3–5 Checkpoints stabil bleibt), kannst du aufhören.

✅ **Methode 2: Maximale Punktzahl**
- Falls die Snake im Durchschnitt bereits sehr viele Punkte erreicht (z. B. 20+ Punkte pro Spiel), ist das Modell stark genug.

✅ **Methode 3: Overfitting verhindern**
- Falls das Modell nur noch sehr spezifische Wege fährt (z. B. immer gleich navigiert, anstatt flexibel zu reagieren), kann das ein Zeichen sein, dass das Modell übertrainiert wurde.

---

### **Fazit: Was du tun solltest**
1. **Nutze `CheckpointCallback`, um Modelle alle 50k Timesteps zu speichern.**  
2. **Überwache das Training mit TensorBoard (`ep_rew_mean`, `policy_loss`, `entropy_loss`).**  
3. **Speichere das beste Modell mit `EvalCallback`, um automatisch die beste Version zu behalten.**  
4. **Falls das Training stagnieren sollte, passe `clip_range` oder `ent_coef` an.**  
5. **Trainiere bis mindestens 250k–500k Timesteps. Falls das Modell gut performt, kannst du stoppen.**  

---

Mit diesem Setup kannst du sicherstellen, dass du **nicht unnötig Ressourcen verschwendest, aber trotzdem ein starkes Modell erhältst**.  

Möchtest du noch eine Anleitung zur **Hyperparameter-Feinjustierung** für dein finales Training?
